File: julia/jl-basic/datatype.jl
# Data type
# define a constant
const MI_CONSTANTE = 100

# use constant. Note: constants are defined at beginning of the file
println(MI_CONSTANTE)

function get_types()

    # Integer
    my_integer = 10
    println("Integer value:", my_integer)
    println("Integer type:", typeof(my_integer))

    # Float
    my_float = 3.14
    println("Float value:", my_float)
    println("Float type:", typeof(my_float))

    # String
    my_string = "Hello, world!"
    println("String value:", my_string)
    println("String type:", typeof(my_string))

    # Boolean
    my_boolean = true
    println("Boolean value:", my_boolean)
    println("Boolean type:", typeof(my_boolean))

    # Array
    my_array = [1, 2, 3, 4, 5]
    println("Array value:", my_array)
    println("Array type:", typeof(my_array))

    # Tuple
    my_tuple = (6, 7, 8, 9, 10)
    println("Tuple value:", my_tuple)
    println("Tuple type:", typeof(my_tuple))

    # Set
    my_set = Set([1, 2, 3])
    println("Set value:", my_set)
    println("Set type:", typeof(my_set))

    # Dictionary
    my_dict = Dict("name" => "John", "age" => 30, "city" => "New York")
    println("Dictionary value:", my_dict)
    println("Dictionary type:", typeof(my_dict))

    # String
    # Multiline

    multiline_string = """
    This is a
    multiline
    string.
    """
    println(multiline_string)

    # Concatenation:

    str1 = "Hello"
    str2 = "World"
    concatenated_str = string(str1, " ", str2)
    println(concatenated_str)

    # Char: In Julia, individual characters are represented using single quotation marks.
    char = 'A'
    println(char)

    # Declaring Variables. Explicit:
    my_explicit_variable::Int64 = 10
    println(my_explicit_variable)

    # Implicit
    my_implicit_variable = 3.14
    println(my_implicit_variable)

    # Group of Declarations:
    var1, var2, var3 = 1, "two", true
    println(var1, " ", var2, " ", var3)

    # Multiple Declarations Inline:
    x = y = z = 0
    println(x, " ", y, " ", z)

    #Check Type:
    my_variable = 10
    println(typeof(my_variable))

    # Casting Data Types
    #=
        Other programming languages arbitrarily chose some way of rounding here (often truncation but sometimes rounding to nearest). Julia doesn't guess and requires you to be explicit. If you want to round, truncate, take a ceiling
    =#

    my_float1 = 3.14656565
    my_int1 = Int64(round(my_float1))
    println(my_int1)
    println(typeof(my_int1))

    #=
        Mutability / Immutability:
        In Julia, variables are mutable by default. However, you can make a variable immutable by using the const keyword or declaring it with the immutable keyword.
    =#

    #=
        Special types:
        Null value:

        In Julia, the null value is represented by nothing. It is used to indicate the absence of a value.

        Pointers & Address:

        Julia is a high-level language that abstracts away direct memory management and pointer operations. Pointers and memory addresses are not typically used in everyday Julia programming.
    =#

end
  
get_types() # Julia doesnt need a main function

File: python/fundamentals/datatype.py
from datetime import date
from datetime import time
from datetime import datetime

"""
    Task Datatype manager
"""
#Task Entry point
def get_types():
    print("*** Types ***")
    
    # Integer
    my_integer = 10
    print("Integer value:", my_integer)
    print("Integer type:", type(my_integer))

    # Float
    my_float = 3.14
    print("Float value:", my_float)
    print("Float type:", type(my_float))

    # String
    my_string = "Hello, world!"
    print("String value:", my_string)
    print("String type:", type(my_string))

    # Boolean
    my_boolean = True
    print("Boolean value:", my_boolean)
    print("Boolean type:", type(my_boolean))

    # List Mutable
    my_list = [1, 2, 3, 4, 5]
    print("List value:", my_list)
    print("List type:", type(my_list))

    # Tuple Inmutalble
    my_tuple = (6, 7, 8, 9, 10)
    print("Tuple value:", my_tuple)
    print("Tuple type:", type(my_tuple))

    # Set
    my_set = {1, 2, 3}
    print("Set value:", my_set)
    print("Set type:", type(my_set))

    # Dictionary
    my_dict = {"name": "John", "age": 30, "city": "New York"}
    print("Dictionary value:", my_dict)
    print("Dictionary type:", type(my_dict))
    
    # Constant
    # Python doesn't support constant
    MY_CONSTANT = "my_constant" # By convention
    print("My constant ", MY_CONSTANT)
    
    my_date = date(2023, 5, 30)
    print(my_date)
    print("date type:", type(my_date))
    
    my_time = time(9, 30, 45)
    print(my_time)
    print("time type:", type(my_time))
    
    my_dt = datetime(2023, 5, 30, 9, 30, 45)
    print(my_dt)
    print("date time type:", type(my_dt))
    
    # String multiline
    ml = """ This is a
    multiline 
    string
    """
    print(ml)
    
    #Concatenation
    str1 = "Hello"
    str2 = "World"
    concatenated_str = str1 + " " + str2
    print(concatenated_str)
    
    #Char: In Python, there is no distinct "char" data type like in some other programming languages. 
    char = 'A'
    print(char)

    # Declaring Variables 
    # Explicit:
    my_explicit_variable: int = 10
    print(my_explicit_variable)
    
    # Implicit:
    my_implicit_variable = 3.14
    print(my_implicit_variable)
    
    # Group of Declarations:
    var1, var2, var3 = 1, "two", True
    print(var1, var2, var3)
    
    # Multiple Declarations Inline
    x = y = z = 0
    print(x, y, z)
    
    # Check Type:
    my_variable = 10
    print(type(my_variable))
    
    # Casting data type
    
    my_int = 10
    my_float = float(my_int)
    print(my_float)
    
"""
Special types:
    Null value:

    In Python, the null value is represented by None. It is often used to indicate the absence of a value or as a default placeholder.

    Pointers & Address:

    Python does not expose pointers and memory addresses to the same extent as some low-level programming languages
 
 """   
    
if __name__ == "__main__":
    get_types()